### PR TITLE
Update cohere2_moe tp_plan

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

Since https://github.com/huggingface/transformers/pull/46115 was merged to be released in 5.9 without https://github.com/huggingface/transformers/pull/45028, the tp_plan was still using the old format. This PR updates it.

### CODE REVIEW

**Review by 3outeille — APPROVED** [review]

File: src/transformers/models/cohere2_moe/configuration_cohere2_moe.py [hunk 1]
```diff
@@ -149,6 +169,29 @@ def __post_init__(self, **kwargs):
             )
 
         super().__post_init__(**kwargs)
+        self._update_sp_plan()
+
+    def _update_sp_plan(self):
+        self.base_model_sp_plan = self.base_model_sp_plan.copy()
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2_moe/configuration_cohere2_moe.py:177]: Can you document this please! 🤗 
```
we have to update the sp plan because some of the layers are MLP other MOE`
```

also can you refacto to use the standard mlp_layer type instead of first k_dense please!

File: src/transformers/models/cohere2_moe/modeling_cohere2_moe.py [hunk 2]
```diff
@@ -87,16 +87,11 @@ def __init__(self, config: Cohere2MoeConfig, intermediate_size=None):
...
+        self.intermediate_size = intermediate_size if intermediate_size is not None else config.intermediate_size
         self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
         self.act_fn = ACT2FN[config.hidden_act]
-        if intermediate_size is not None:
-            self.intermediate_size = intermediate_size
-            self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
-            self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
-            self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2_moe/modeling_cohere2_moe.py:99]: lol

File: tests/models/cohere2_moe/test_modeling_cohere2_moe.py [hunk 3]
```diff
@@ -57,6 +57,7 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.layer_types = ["full_attention", "sliding_attention"]
         self.first_k_dense_replace = 1  # first layer will be MLP, 2nd will be MoE
+        self.logit_scale = 1.0  # needed for `test_training_overfit` - otherwise the loss does not go down fast enough
```
Comment (ArthurZucker) [inline tests/models/cohere2_moe/test_modeling_cohere2_moe.py:60]: can you put it in `CausalLMModelTester` pretty sure other models might hav ethis issue no?

**Review by ArthurZucker — APPROVED** [review]
lgtm

File: tests/models/cohere2_moe/test_modeling_cohere2_moe.py (hunk 3 above)
Comment (Cyrilvallez) [reply tests/models/cohere2_moe/test_modeling_cohere2_moe.py:60]: Other cohere models do not need it for the loss to go down... So I'd rather keep this isolated to cohere2_moe!
